src/singleton.cr

```diff
@@ -0,0 +1,39 @@
+# The Singleton module implements the Singleton pattern.
+#
+### Usage
+#
+# ```
+# require "singleton"
+#
+# class Foo
+#   include Singleton
+# end
+#
+# foo = Foo.instance
+# ```
+
+module Singleton
+  macro included
+    @@mutex = Mutex.new
+
+    # Instantiate the object if it doesn't exist.  Always returns the same object.
+    def self.instance
+      @@instance ||= @@mutex.synchronize do
+        @@instance || new
```

Comment:
Assignment here is needed. Otherwise there is a small window for a race condition and initialize the object twice.